### ISSUE TITLE
Plasma 5 support

### ISSUE THREAD

**Issue by willprice** [faho/kwin-tiling#30]
Hi,

I've tried installing the script under the new Plasma 5 desktop but can't seem to get it working. I appreciate this isn't a very helpful bug report as it contains no information whatsoever, but I'm not sure what is relevant. I'd be happy to help with alpha/beta testing on plasma 5.

**Comment by faho** [comment]
I've found the cause for the configuration menu issues, it's now https://bugs.kde.org/show_bug.cgi?id=347176.

**Comment by faho** [comment]
The config thing has been fixed by fcc931cccf25cb2d9bef6c182a3a6dc503db4940,
the shortcuts and clientlist thing remain, though the latter is pretty well worked around.

**Comment by faho** [comment]
I just moved the plasma5 branch to master and the old master to plasma1.

**Comment by tidux** [comment]
The extension is broken on Fedora 22 with Plasma 5.3.2.  The error is "Plugin does not provide configuration file in expected location".

**Comment by faho** [comment]
What does KWin print when run in a terminal (`kwin_x11 --replace`)?

Is this after fcc931cccf25cb2d9bef6c182a3a6dc503db4940?

**Comment by tidux** [comment]
This was git master about half an hour before I filed the issue.
```
$ kwin_x11 --replace
kwin_core: Extensions: shape: 0x "11"  composite: 0x "4"  render: 0x "b"  fixes: 0x "50"  randr: 0x "14"  sync: 0x "31"  damage: 0x  "11" 
    
Constructing a KPluginInfo object from old style JSON. Please use kcoreaddons_desktop_to_json() for "/usr/lib64/qt5/plugins/org.kde.kdecoration2/breezedecoration.so" instead of kservice_desktop_to_json() in your CMake code.
    Constructing a KPluginInfo object from old style JSON. Please use kcoreaddons_desktop_to_json() for "/usr/lib64/qt5/plugins/org.kde.kdecoration2/kwin5_aurorae.so" instead of kservice_desktop_to_json() in your CMake code.
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 38743949
kwin_core: User timestamp, final: 'ID: 44040207 ;WMCLASS: "plasmashell" : "plasmashell" ;Caption: "Desktop — Plasma" ' : 38743949
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 41678101
kwin_core: User timestamp, final: 'ID: 81788949 ;WMCLASS: "okular" : "okular" ;Caption: "Eclipse Phase Quick-Start Rules – Okular" ' : 41678101
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 4294967295
kwin_core: User timestamp, final: 'ID: 18874378 ;WMCLASS: "urxvt" : "urxvt256c-ml" ;Caption: "urxvt256c-ml" ' : 4294967295
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 52142044
kwin_core: User timestamp, final: 'ID: 92274763 ;WMCLASS: "libreoffice-calc" : "libreoffice" ;Caption: "_Character_Builder_0.3.2.1.xlsx - LibreOffice Calc" ' : 52142044
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 4294967295
kwin_core: User timestamp, final: 'ID: 4194314 ;WMCLASS: "urxvt" : "urxvt256c-ml" ;Caption: "HACKING THE GIBSON on daemonhost" ' : 4294967295
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 53269449
kwin_core: User timestamp, final: 'ID: 75497498 ;WMCLASS: "firefox" : "navigator" ;Caption: "Plasma 5 support · Issue #30 · faho/kwin-tiling - Vimperator" ' : 53269449
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 4294967295
kwin_core: User timestamp, final: 'ID: 67108874 ;WMCLASS: "urxvt" : "urxvt256c-ml" ;Caption: "urxvt256c-ml <2>‎" ' : 4294967295
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: User timestamp, ASN: 52087351
kwin_core: User timestamp, final: 'ID: 44040214 ;WMCLASS: "plasmashell" : "plasmashell" ;Caption: "Plasma" ' : 52087351
kwin_core: Activation: No client active, allowing
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
QXcbConnection: XCB error: 8 (BadMatch), sequence: 1707, resource id: 67108874, major code: 42 (SetInputFocus), minor code: 0
kwin_core: Initializing OpenGL compositing
kwin_core: Choosing GLXFBConfig 0x77 X visual 0xb0 depth 24 RGBA 8:8:8:0 ZS 0:0
OpenGL vendor string:                   Intel Open Source Technology Center
OpenGL renderer string:                 Mesa DRI Intel(R) Ivybridge Mobile 
OpenGL version string:                  3.3 (Core Profile) Mesa 10.6.3 (git-ccef890)
OpenGL shading language version string: 3.30
Driver:                                 Intel
GPU class:                              IvyBridge
OpenGL version:                         3.3
GLSL version:                           3.30
Mesa version:                           10.6.3
Linux kernel version:                   4.1.6
Requires strict binding:                yes
GLSL shaders:                           yes
Texture NPOT support:                   yes
Virtual Machine:                        no
kwin_core: Color correction: false
kwin_core: OpenGL 2 compositing successfully initialized
kwin_core: Vertical Refresh rate  60 Hz ( "primary screen" )
QXcbConnection: XCB error: 9 (BadDrawable), sequence: 1949, resource id: 0, major code: 14 (GetGeometry), minor code: 0
QXcbConnection: XCB error: 9 (BadDrawable), sequence: 1959, resource id: 0, major code: 14 (GetGeometry), minor code: 0
QXcbConnection: XCB error: 9 (BadDrawable), sequence: 1960, resource id: 0, major code: 14 (GetGeometry), minor code: 0
QXcbConnection: XCB error: 9 (BadDrawable), sequence: 1961, resource id: 0, major code: 14 (GetGeometry), minor code: 0
kwin_core: screens:  1 desktops:  4
kwin_core: Done.
kwin_core: Successfully loaded built-in effect:  "blur"
kwin_core: Successfully loaded built-in effect:  "contrast"
kwin_core: Successfully loaded built-in effect:  "dashboard"
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_windowaperture"
kwin_core: Session path: "/org/freedesktop/login1/session/_35"
kwin_core: Successfully loaded built-in effect:  "desktopgrid"
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_fade"
kwin_core: Successfully loaded built-in effect:  "highlightwindow"
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_maximize"
kwin_core: Successfully loaded built-in effect:  "kscreen"
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_dialogparent"
kwin_core: Successfully loaded built-in effect:  "logout"
kwineffects: Incorrect Property state, immediate stop:  7471104
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_login"
kwin_core: Successfully loaded built-in effect:  "minimizeanimation"
kwin_core: Successfully loaded scripted effect:  "kwin4_effect_translucency"
kwin_core: Successfully loaded built-in effect:  "presentwindows"
kwin_core: Successfully loaded built-in effect:  "screenedge"
kwin_core: Successfully loaded built-in effect:  "screenshot"
kwin_core: Successfully loaded built-in effect:  "slide"
kwin_core: Successfully loaded built-in effect:  "slidingpopups"
kwin_core: Successfully loaded built-in effect:  "startupfeedback"
kwin_core: Successfully loaded built-in effect:  "zoom"
```

**Comment by faho** [comment]
There's nothing in there related to the script, even though we print quite a few debug lines.

How did you install it? Note that the README was broken for a while - it needs to be via `plasmapkg2 --type kwinscript -i /path/to/script`. Or maybe Fedora is doing something weird in its packaging.

**Comment by tidux** [comment]
I did install with that `plasmapkg2` syntax.  It's quite possible Fedora is doing something weird, or something changed in Plasma or kwin recently.

**Comment by faho** [comment]
Seems I need to get a Fedora VM again. Anything special you have to do to get Plasma 5?

I doubt it's Plasma/KWin since I'm running 5.4 (and had run 5.3.2 when that was current).

**Comment by tidux** [comment]
Just install the KDE spin and sudo dnf update.

On September 1, 2015 6:59:38 PM CDT, Fabian Homborg <notifications@github.com> wrote:
>Seems I need to get a Fedora VM again. Anything special you have to do
>to get Plasma 5?
>
>I doubt it's Plasma/KWin since I'm running 5.4 (and had run 5.3.2 when
>that was current).
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/faho/kwin-tiling/issues/30#issuecomment-136896150

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by faho** [comment]
Sorry it's taking so long, I'm having trouble getting Fedora to work properly (even before installing the script).

**Comment by v3ss0n** [comment]
Tested on Sabayon Linux , Plasma 5.4 and its working good! thank you so much ! 
Keep up the goodjob!

**Comment by v3ss0n** [comment]
Shortcut seems (meta + f11 + shift) to do nothing, no shortcut support yet?

**Comment by faho** [comment]
Nope. It's currently impossible once you use qml, and if we didn't use qml we wouldn't have Qt.include - i.e. all code would have to be in one file. One 3000 line file.

I've reported it upstream a while ago, but they've got other things on their plate right now.

**Comment by v3ss0n** [comment]
may be we can use node + requirejs and generate them , without qml? my 2 cents.

**Comment by faho** [comment]
Nah - if I really wanted to get rid of qml, I'd use a buildscript that mainly invoked `cat`. But I've always had plans to use qml for _more_ (like a popup to show which layout is being loaded), not less, and P5 is still not quite "finished" (which is also why I haven't made a P5-only release and uploaded it to kde-look.org) and shortcuts aren't critical, so... let's just wait until upstream fixes it.

**Comment by v3ss0n** [comment]
yes that make sense :D lets wait for it. For now if i want to switch default layout can i do at settings?

**Comment by faho** [comment]
> For now if i want to switch default layout can i do at settings?

Should be possible, yeah (though the interface isn't great because I can't figure out how to make one that's flexible enough). That last textfield in the "Common Options" tab takes text in the form "desktop:layout", with multiple of those separated with a ",". A negative desktop number disables tiling by default on that desktop, and a desktop of "0" indicates that this is the default for all desktops that aren't mentioned otherwise.

**Comment by v3ss0n** [comment]
so same as syntax in the last screenshot? I gonna try it! thanks.

**Comment by rafasc** [comment]
still no fix to keyboard shortcuts?

**Comment by faho** [comment]
Unfortunately not, and I don't think it's going to be in KWin 5.5 either.

**Comment by v3ss0n** [comment]
that gonna be long long long wait.

**Comment by marccollin** [comment]
i use plasma 5.4.3, i see the scrit p in the kwin script.

i try to enable it but get this message "The shared library was not found."

**Comment by faho** [comment]
>i try to enable it but get this message "The shared library was not found."

That's been known to show up when you click the config menu, but the script should be active regardless (once you restart kwin). As for the config menu - that should be solved by fcc931cccf25cb2d9bef6c182a3a6dc503db4940, so you should try that. If it's not, can you tell me where "kcm_kwin4_genericscripted.so" is? It should be included in your kwin package.

**Comment by marccollin** [comment]
locate tiling
/home/collinm/.local/share/kservices5/kwin-script-kwin-script-tiling.desktop
/home/collinm/.local/share/kwin/scripts/kwin-script-tiling/metadata.desktop

i modified

/home/collinm/.local/share/kservices5/kwin-script-kwin-script-tiling.desktop
/home/collinm/.local/share/kwin/scripts/kwin-script-tiling/metadata.desktop

to put this value
X-KDE-Library=/usr/lib64/qt5/plugins/kwin/effects/configs/kcm_kwin4_genericscripted

i get same result

**Comment by marccollin** [comment]
seem like don't use the path i specified

**Comment by Hi-Angel** [comment]
You said that shortcuts are not that critical — does that mean I can configure, e.g., the changing tiling layout shortcut somehow else?

**Comment by faho** [comment]
> the changing tiling layout shortcut somehow else?

No. You can change the layout in the configuration (or hardcode it in the actual code), but shortcuts currently (AFAIK still) just do not work. Any change in layout requires a KWin restart.

They're not critical in that the script is basically usable without them. It's secondary functionality.

**Comment by Hi-Angel** [comment]
If anybody want to vote for, I reported a [feature request to make tiling upstream](https://bugs.kde.org/show_bug.cgi?id=369068).

**Comment by steverweber** [comment]
Thanks @Hi-Angel

**Comment by v3ss0n** [comment]
yes lets vot for this , lets get this plugin mainstream!

**Comment by v3ss0n** [comment]
voted @Hi-Angel

**Comment by faho** [comment]
While I appreciate your enthusiasm, I don't think that's gonna happen.

KWin actually used to have tiling. It was written for GSoC and quite minimalistic and generally bad. It was removed because none of the kwin devs used it or cared for it. They recommended that someone write a script instead, and that's how this thing started.

Now, if KWin would incorporate this script, they'd still need to fix the API issues - which they obviously haven't gotten to. If they added it to the core code instead, they'd have to write quite a bunch of code. Either way, there's no real advantage for the kwin devs.

**Comment by v3ss0n** [comment]
you version of tiling is much better and stable. I had use that official before. So , let us try , may be a lot of votes can change something.

**Comment by Hi-Angel** [comment]
Honestly, I rather want to stir them up. They would think for a moment "What do we need for this to work?", and then they would remember the ancient hotkeys bug, since 2014. Then they can't ask "why not to keep it separate of upstream", because the answer is obviously "The API is broken". It's not cool to have bugs appearing, depending on an old one, thus the hotkeys bug more likely to get fixed, then without the report.

**Comment by ohjames** [comment]
I fixed keyboard shortcuts in QML now ;) Could anyone tell me if it's possible to remove/update a kwinscript? I've tried the -u/-r options to plasmapkg2 but they seem broken. How do you guys handle it when developing kwin-tiling?

**Comment by v3ss0n** [comment]
Hiw.Can you share us the fix??

**Comment by Hi-Angel** [comment]
@v3ss0n it's here https://bugs.kde.org/show_bug.cgi?id=340225#c22

**Comment by faho** [comment]
>How do you guys handle it when developing kwin-tiling?

I actually symlink the script install dir to the git repo.

**Comment by ohjames** [comment]
I see, symlink and `kwin --replace` to reload. I'll try and get plasmapkg2 to work properly then. After that I can work on some of your other issues, if you could let me know which are most important I'll get to them first.

**Comment by faho** [comment]
>After that I can work on some of your other issues, if you could let me know which are most important I'll get to them first.

Woohoo, thanks!

I'm assuming you mean upstream kwin issues. If so, the one that bugs me the most is https://bugs.kde.org/show_bug.cgi?id=341195, which is #6 here (under certain circumstances, we can cause clients to not update graphically anymore).

After that, there's https://bugs.kde.org/show_bug.cgi?id=340125, which might be much smaller and is more of a minor annoyance for developing this thing.

**Comment by ohjames** [comment]
https://bugs.kde.org/show_bug.cgi?id=341195 could be difficult to fix, but I can easily fix https://bugs.kde.org/show_bug.cgi?id=340125

**Comment by ohjames** [comment]
@faho btw could you explain how inheritance is broken? You mean as in Ecmascript inheritance via `childClass.prototype = Object.create(parentClass.prototype)` (or maybe QML doesn't support `Object.create` and you need the worse: `childClass.prototype = new parentClass`? Neither of those?

**Comment by faho** [comment]
@ohjames: That one was a wrong assumption on my part. KWin 5 is a bit stricter when it comes to includes, and I was doing the latter when I hadn't included the parent directly.

**Comment by ohjames** [comment]
I see. Do you know if ES5.1 methods like Object.create are available from declarativescript? I'm pretty sure their engine uses a recentish version of v8 and should do, but I couldn't find `Object.create` in kwin-tiling's source so I wasn't sure if it is possible. BTW I've responded to all of the code review items so the registerShortcut patch should hopefully be pulled soon. My clone in github has it in case you want to try it out.

**Comment by faho** [comment]
>Do you know if ES5.1 methods like Object.create are available from declarativescript?

Nope. To be honest that's the first I've heard of them. I'm not much of a web guy.

>BTW I've responded to all of the code review items so the registerShortcut patch should hopefully be pulled soon.

Very nice!

**Comment by ohjames** [comment]
I've fixed https://bugs.kde.org/show_bug.cgi?id=340125 also now. Without it you can watch for new clients being added and cache them that way, but not get the existing list of clients?

**Comment by faho** [comment]
>Without it you can watch for new clients being added and cache them that way, but not get the existing list of clients?

Yeah. Which is why it's annoying for development. When you restart kwin, it won't automatically tile existing clients.

For normal use, it's not really important.

**Comment by ohjames** [comment]
I find the feature really useful, I'm using it in [kwin window switch](https://github.com/ohjames/kwin-window-switch) already to simplify switching between windows... without it I'd have to watch for windows to open.

**Comment by faho** [comment]
So, current git master basically works on Plasma 5. It's installable since a kind soul uploaded it to store.kde.org.

My current plan is to clean it up a bit again (I've already fixed some shortcut functions that were broken for P5), find some way to configure excludes - #64, and then make a first actual plasma 5 release. It should be widespread enough now.

**Comment by v3ss0n** [comment]
thank you very much for the great work!

**Comment by pdf** [comment]
After fighting to install this for a while, I've discovered that support for installing kwin script packages is currently broken in 5.8 (at least on Arch, 5.8.5/5.30.0), and won't be fixed until 5.9 is released by the looks: https://bugs.kde.org/show_bug.cgi?id=374766

I tried manually stuffing things into `~/.local/share/kwin/scripts` and `~/.local/share/kservices5`, but whilst I could get the plugin to show up in the list of scripts, edit the config, etc, the script does not appear to run.

BTW, plasmapkg2 is apparently deprecated and replaced with kpackagetool5 (on my Arch system, plasmapkg2 appears to be a thin wrapper that just calls kpackagetool5).

**Comment by faho** [comment]
This has been fixed for a while now.
